packages/babel-plugin-transform-es2015-destructuring/src/index.js

```diff
@@ -491,16 +491,27 @@ export default function ({ types: t }) {
         }
 
         const nodesOut = [];
-        for (const node of nodes) {
-          const tail = nodesOut[nodesOut.length - 1];
-          if (
-            tail && t.isVariableDeclaration(tail) && t.isVariableDeclaration(node) &&
-            tail.kind === node.kind
-          ) {
-            // Create a single compound let/var rather than many.
-            tail.declarations.push(...node.declarations);
-          } else {
-            nodesOut.push(node);
+
+        // Group all declarations to a single node to avoid
+        // wrong transformation in the `for` initialization.
+        // See https://github.com/babel/babel/issues/4893
+        if (t.isForStatement(parent)) {
```

Comment:
Great idea, thanks! New commit: https://github.com/babel/babel/pull/5763/commits/65d47de6cb13e97cf28a9886d5ed66d6555d63da. Some existing tests were updated given the new merge behavior. Please let me know what you think.